jujusvg.go

```diff
@@ -10,11 +12,51 @@ import (
 	"gopkg.in/juju/charm.v5-unstable"
 )
 
-// NewFromBundle returns a new Canvas that can be used
+func NewFromBundleEmbedIcons(b *charm.BundleData, iconURL func(*charm.Reference) string) (*Canvas, error) {
+	alreadyFetched := make(map[string]bool)
+	icons := make(map[string]string)
+	for _, serviceData := range b.Services {
+		charmId, err := charm.ParseReference(serviceData.Charm)
+		if err != nil {
+			return nil, errgo.Notef(err, "cannot parse charm %q", serviceData.Charm)
+		}
+		if _, ok := alreadyFetched[charmId.Path()]; ok == false {
+			alreadyFetched[charmId.Path()] = true
+			icon, err := fetchIcon(iconURL(charmId))
+			if err != nil {
+				return nil, err
+			}
+			icons[charmId.Path()] = icon
+		}
+	}
+	return NewFromBundleWithMap(b, iconURL, icons)
+}
+
+func fetchIcon(url string) (string, error) {
+	resp, err := http.Get(url)
```

Comment:
if !alreadyFetched[charmId.Path()] { would do the same thing (it's false by default) and be more readable.